[dnf5] doc: Document RepoSet and RepoQuery

dmach opened the pull request.

Commit:
doc: Document RepoSet and RepoQuery

j-mracek commented on include/libdnf/rpm/package_set.hpp:
```diff
@@ -70,45 +69,98 @@ class PackageSet {
-    friend libdnf::Goal;
-    friend libdnf::Swdb;
+    friend class PackageSetIterator;
+    friend class PackageQuery;
```
I propose to change `class PackageQuery;` to `struct PackageQuery;`. Anyway we have to change it for C bindings.
I suggest to keep`friend PackageQuery;` is a better idea.

dmach replied:
As discussed in the meeting earlier today, I'd prefer to stick with `class` and if we decide to support C API, then mass-rename `class` to `struct` everywhere in the code. AFAIK even `friend struct <Class>` should work fine and is straightforward to replace.

Conan-Kudo replied:
Is there a reason we can't just have SWIG generate C bindings?

Commit:
Move set operation methods from PackageQuery to PackageSet

It is not possible to use only the operators, because they are not
exposed in Ruby bindings properly.

Commit:
Rename PackageQuery::filter_repoid() to filter_repo_id()

j-mracek commented on include/libdnf/rpm/package_query.hpp:
```diff
@@ -77,285 +77,495 @@ class PackageQuery : public PackageSet {
         const std::vector<std::string> & patterns, libdnf::sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
 
-    /// cmp_type could be only libdnf::sack::QueryCmp::EQ, NEQ, GLOB, NOT_GLOB, IEXACT, NOT_IEXACT, ICONTAINS, NOT_ICONTAINS, IGLOB, NOT_IGLOB, CONTAINS, NOT_CONTAINS.
+    /// Filter packages by their `name-[epoch:]version-release.arch` based on values set in the `nevra` object.
```
What about to replace `name-[epoch:]version-release.arch` by `name`, `epoch`, `version`, `release`, and `arch`?

dmach replied:
I intentionally used this to indicate what's the input format.
Individual values do not describe the input format at all.

j-mracek replied:
I believe that the input format is not relevant at all. You can set each string separately in nevra. You even don't need to parse string or parse string in NEVRA format.

dmach replied:
Right, this wasn't entirely correct because it's referring to nevra string rather than object attributes.
I've rephrased the doctext.

j-mracek commented on include/libdnf/rpm/package_query.hpp:
```diff
@@ -77,285 +77,495 @@ class PackageQuery : public PackageSet {
+        const libdnf::rpm::Nevra & nevra, libdnf::sack::QueryCmp cmp_type = libdnf::sack::QueryCmp::EQ);
 
-    /// cmp_type could be only libdnf::sack::QueryCmp::EQ, NEQ, GLOB, NOT_GLOB, IEXACT, NOT_IEXACT, ICONTAINS, NOT_ICONTAINS, IGLOB, NOT_IGLOB, CONTAINS, NOT_CONTAINS.
+    /// Filter packages by their `name-[epoch:]version-release.arch` attributes of the packages in the `package_set`.
```
What about to replace `name-[epoch:]version-release.arch` by `name`, `epoch`, `version`, `release`, and `arch` attributes?

dmach replied:
I intentionally used this to indicate what's the input format.
Individual values do not describe the input format at all.

j-mracek commented on include/libdnf/rpm/package_query.hpp:
```diff
@@ -364,40 +574,41 @@ class PackageQuery : public PackageSet {
-    /// @version 1.0.0
-    /// @param limit Default value is 1
-    /// @returns Self
+    /// Filter packages by grouping them by `name` and `arch` and keeping up to `limit` number of them from each group.
```
What about?
```/// Filter packages by grouping them by `name`, `arch` and version and keeping up to `limit` number of them from each group.```
Without the version it looks like that it returns number of architectures

dmach replied:
I don't think the sentence you have proposed is correct.
Version (and Epoch and Release) is not part of the grouping.
Or did I miss anything? In that case, could you provide several examples how the filter works?

j-mracek replied:
I suggest that the solution is somewhere in the middle. It groups according `name` and `arch`, And inside each group it groups according to versions (EVR).  Then for each `name` and `arch` group it keeps number of EVR groups according to `limit` value.

lukash replied:
@j-mracek then FWIW @dmach's formulation seems to be the correct one.

j-mracek replied:
In case when multiple packages with the same NEVRA are present, then the formulation is misleading.
`A-1-1.noarch`
`A-1-1.noarch`
`A-1-2.noarch`
`A-1-2.noarch`
With `limit=1` it will return `A-1-2.noarch` and `A-1-2.noarch`. 
From `and keeping up to `limit` number of them from each group` I expect to return only one package from `name.arch` group whem `limit=1`.

j-mracek replied:
Suggesting a minor change in `and keeping up to limit number of them from each group` to `and keeping up to limit number of versions from each group`

dmach replied:
I did my best to rephrase the docstring to make it more clear. Please take a look.

lukash commented on include/libdnf/rpm/package_set.hpp:
```diff
@@ -74,6 +74,28 @@ class PackageSet {
+    /// Throw UsedDifferentSack exceptin when other has a different PackageSack from this
+    /// @replaces libdnf/hy-query.h:function:hy_query_difference(HyQuery q, HyQuery other)
+    /// @replaces libdnf/sack/query.hpp:method:queryDifference(Query & other)
+    void difference(const PackageSet & other) { *this -= other; }
```
We already have the set operators directly above. Can we only have a single method to do these for consistency (of the code using this API) sake?

dmach replied:
I decided to keep both for now.
Ruby doesn't support the operators, they're ignored.
Moving from operators to named methods would be probably too big for this PR.

lukash commented on include/libdnf/rpm/package_query.hpp:
```diff
@@ -77,285 +77,495 @@ class PackageQuery : public PackageSet {
 
-    /// cmp_type could be only libdnf::sack::QueryCmp::EQ, NEQ, GLOB, NOT_GLOB.
-    PackageQuery & filter_epoch(
+    /// Filter packages by their `name-[epoch:]version-release.arch`. The following matches are tolerant to omited 0 epoch: `EQ`, `NEQ`, `GT`, `GTE`, `LT`, `LTE`.
```
Typo: omitted

lukash commented on include/libdnf/rpm/package_query.hpp:
```diff
@@ -77,285 +77,495 @@ class PackageQuery : public PackageSet {
+    /// @param cmp              A comparison (match) operator, defaults to `QueryCmp::EQ`.
+    ///                         Supported values: `EQ`, `NEQ`.
+    /// @since 5.0
+    /// cmp_type could be only libdnf::sack::QueryCmp::EQ, NEQ
```
A forgotten original doc line.

Commit:
Rename PackageQuery::filter_latest() to filter_latest_evr()

This makes more clear that the filter runs on evrs
and can return multiple packages for each.

Also make some tweaks to CHANGES-DNF-5.md.

j-mracek commented on include/libdnf/rpm/package_query.hpp:
```diff
@@ -383,7 +383,7 @@ class PackageQuery : public PackageSet {
     /// @replaces libdnf/sack/query.hpp:method:addFilter(int keyname, int cmp_type, int match) - cmp_type = HY_PKG_LATEST
     /// @replaces libdnf/sack/query.hpp:method:addFilter(int keyname, int cmp_type, int match) - cmp_type = HY_PKG_LATEST_PER_ARCH
-    PackageQuery & filter_latest(int limit = 1);
+    PackageQuery & filter_latest_evr(int limit = 1);
```
Are you sure that the name `filter_latest_evr` is more informative? It replaces `HY_PKG_LATEST_PER_ARCH` and obsoletes `HY_PKG_LATEST`.

dmach replied:
Let's keep it this way to give users a hint that it's **not** the latest package but that it works with groups of nevras.
A more correct method name would be something like filter_latest_evr_per_na(), but that would be probably too much :)

j-mracek commented:
LGTM
